### Agent review — round 1

**Reviewer:** independent `reviewer` subagent (Claude Sonnet 4.6), no authorship of this branch
**Scope:** `90d1f5a3ed0bcb435830be0900d3363637ba9b66..42e66a635c11a5f726df658dbdc051cedfd8c267`
**Verdict:** **changes requested** — 1 blocker, 2 major.

The rebase itself is clean: this commit does **not** duplicate or conflict with main's `_on_conversation_name_written` done-callback or `_NAME_PERSIST_MAX_PASSES`, and the flush-budget change composes correctly with main's bounded-retry comment. The word cut and the flush budget are correct and well-pinned. The band change, however, drops the reserve on one real configuration, and the three new band tests do not pin any of the band behaviour they were written for.

| ID | Severity | Area | Finding |
|---|---|---|---|
| F1 | **BLOCKER** | band | The name's reserve is silently discarded when the name is the only right-hand segment, so the title's column moves with its own length |
| F2 | **MAJOR** | tests | All three new band tests pass unmodified against `main` — they pin none of fix 1 or fix 2 |
| F3 | **MAJOR** | band | F1's lost reserve changes which siblings the ladder sheds, so a longer title costs a segment a shorter one keeps |
| F4 | MINOR | tests | `test_a_slow_but_real_write_keeps_its_title` passes against main's hardcoded `2.0`; only its sibling discriminates |
| F5 | NIT | naming | `_cut_on_a_word` private alias is an unused indirection |

---

#### F1 — BLOCKER. The reserve is dropped when the name is the only right-hand segment

`local_operator/tui/widgets/status_line.py:1704-1713`. The slack is emitted inside the join loop, guarded by `if index:`:

```python
for index, (icon, text, style) in enumerate(parts):
    if index:                                            # <-- never true for index 0
        if index == len(parts) - 1 and self._name_slack:
            right.append(" " * self._name_slack, style=dim)
        right.append(f" {_SEP_RIGHT} ", style=seam)
```

When the name is the **only** element of `parts`, it is at `index == 0`, `if index:` is false, and the slack — correctly computed on line 1692 — is never painted. The task brief asks specifically whether `index == len(parts) - 1` "genuinely identifies the name segment in every configuration"; this is the configuration where it does not, because the guard above it swallows the case.

Reproduced against the real `StatusLine` (`_right_text` called directly, `NAME_CELLS` reserve):

```
name alone   -> right.plain='Fix boot'                                  slack=32  EMITTED=False
alarm + name -> right.plain='!                                 ‹ Fix boot' slack=32  EMITTED=True
```

The consequence is precisely the defect `NAME_CELLS` exists to prevent — the title's column becomes a function of the title's text. Same band, width 100, name as sole right-hand segment:

```
=== MAIN 90d1f5a ===
  'a'                                        title starts col= 72   row=73 cells
  'Fix boot'                                 title starts col= 72   row=80
  'Bulk export the invoice columns'          title starts col= 72   row=96
  'Fix unnamed sessions from image openers'  title starts col= 72   row=98
  distinct title columns: [72]  -> STABLE

=== HEAD 42e66a6 ===
  'a'                                        title starts col= 99   row=100
  'Fix boot'                                 title starts col= 92   row=100
  'Bulk export the invoice columns'          title starts col= 76   row=100
  'Fix unnamed sessions from image openers'  title starts col= 74   row=100
  distinct title columns: [74, 76, 92, 99]  -> MOVES WITH TITLE
```

The row does reach the edge, but by right-aligning the group **by its ink** — the alternative the PR description explicitly rejects ("Measuring the right group by its ink … walks the alarm column 74->91/111/151"). On this configuration the commit implements that rejected alternative, for the name itself.

This is not a synthetic band. `format_jobs(0)`, `format_agents(0)` and `format_context_usage(0, …)` all return `""`, cost is absent until the first billed call, and `duration` is suppressed while `elapsed == 0` — so a **freshly opened session, at the instant `provisional_title` names it from the opener**, has exactly one right-hand segment: the name. That is the first band the user sees, and the title visibly jumps left as the generated title replaces the provisional excerpt:

```
=== HEAD, w=72, fresh session ===
provisional |◆ kimi-k2-thinking › ▴ high › ⌂ local-operator       The /resume picker…|
generated   |◆ kimi-k2-thinking › ▴ high › ⌂ local-operator                  Fix boot|
```

Fix in one line: hoist the slack append out of the `if index:` guard — emit it before the seam when `index`, and before the text when `index == 0` and the name is the sole part.

#### F2 — MAJOR. The three new band tests pass against `main` unchanged

Checked by the strongest available method: reverting **all four source files** to `90d1f5a` while keeping this commit's tests, then running them.

```console
$ git checkout 90d1f5a -- local_operator/    # tests stay at HEAD
$ pytest tests/unit/tui/test_status_line.py tests/unit/tui/test_terminal_title.py \
         tests/unit/session/test_conversation_name_persistence.py -q
FAILED test_terminal_title.py::test_a_renamed_conversation_reaches_the_tab_cut_on_a_word
FAILED test_conversation_name_persistence.py::test_teardown_costs_one_budget_not_two
FAILED test_conversation_name_persistence.py::test_a_slow_but_real_write_keeps_its_title
3 failed, 109 passed
```

The word cut (fix 3) and the flush budget (fix 4) are genuinely pinned. **Every band test passes on main**, including all three added here:

```console
$ pytest tests/unit/tui/test_status_line.py -k "brief_title or seam_stays or double_width" -v
test_a_brief_title_leaves_no_dead_run_at_the_bands_right_edge PASSED
test_the_titles_seam_stays_tight_against_the_title            PASSED
test_a_double_width_title_never_overflows_the_band            PASSED
```

Two distinct problems:

- `test_a_brief_title_leaves_no_dead_run_at_the_bands_right_edge` asserts `row == row.rstrip()`. Main satisfies that trivially, because main's `_compose` **calls `rstrip()`** — the very line this commit removes. The test cannot distinguish "the row reaches the edge" from "the row was trimmed"; it never asserts `cell_len(row) == width`, which is the actual property and the one the PR's before/after table quotes (107/120 -> 120/120).
- `test_a_double_width_title_never_overflows_the_band` finds **zero** over-width rows on main across the same sweep the PR reports 95 for:

  ```console
  $ # 4 CJK/Hangul titles x 4 ladder variants x alarm on/off x widths 40-200
  MAIN 90d1f5a -> over-width rows: 0
  HEAD 42e66a6 -> over-width rows: 0
  ```

  Main pads with `str.ljust`, and `_compose` then `rstrip`s the padding away before the row is measured — so the character/cell mismatch never reaches the output. The overflow the test describes is reachable from `rjust` (which the PR tried and rejected), not from the shipped `main`. The docstring's "592 rows over their width" and the PR's "95" are measurements of an intermediate experiment, not of `90d1f5a`.

The band's four invariants **do** hold on the swept full band at HEAD (I verified 0 over-width, 0 non-flush, 0 orphaned seams, 0 alarm-column drift across widths 60-180) — the code is right there. The gap is that nothing in the suite would catch a revert, and no test covers the single-right-segment shape that F1 breaks.

Suggested: assert `cell_len(row) == width` whenever the name is shown, and add a case where the name is the only right-hand segment.

#### F3 — MAJOR. The lost reserve changes what the ladder sheds

Follows from F1 and is worth its own ID because it breaks a different invariant — the one `_walk` documents as "a long title never sheds a neighbour that a short one would have kept". With the slack unpainted, `right.plain` is shorter than the box `_walk` reserved, so the fit test measures ink rather than the box:

```console
$ # band whose right group is the name alone; short vs long title
MAIN 90d1f5a: 0 widths where the title's LENGTH changed the ladder
HEAD 42e66a6: 40 widths (50-67, 99-120) where it did
```

Rendered at width 55 — the short title keeps the name and loses `▴ high`; the long title keeps `▴ high` and loses the name:

```
=== MAIN ===
short |◆ kimi-k2-thinking › ▴ high › ⌂ local-operator|   effort=True  name=False
long  |◆ kimi-k2-thinking › ▴ high › ⌂ local-operator|   effort=True  name=False

=== HEAD ===
short |◆ kimi-k2-thinking › ⌂ local-operator          Fix boot|  effort=False name=True
long  |◆ kimi-k2-thinking › ▴ high › ⌂ local-operator|          effort=True  name=False
```

A model rewriting the title mid-session makes an unrelated segment appear and disappear. Fixing F1 fixes this too — they are one defect, reported separately because they violate two separately documented properties. Note the fully populated swept band shows **0** divergence, which is why the existing ladder tests miss it: they never exercise the shape where the name stands alone.

#### F4 — MINOR. One flush test does not discriminate

`test_a_slow_but_real_write_keeps_its_title` passes against main's actual code (two hardcoded `timeout=2.0` halves), because its 0.45 s append fits inside a single 2.0 s half:

```console
$ # both waits restored to main's literal 2.0
test_a_slow_but_real_write_keeps_its_title PASSED
```

It fails on main only via the `monkeypatch.setattr(session_mod, "_NAME_FLUSH_TIMEOUT_S", …)` `AttributeError` — i.e. because the constant is absent, not because the behaviour differs. `test_teardown_costs_one_budget_not_two` **is** discriminating (it fails at 0.80 s when both halves are given a full budget each, and at 4.00 s against main's literal), so the pair is adequate overall; this one just does not carry its own weight. Widening its append past the per-half budget would make it real.

#### F5 — NIT. `_cut_on_a_word` alias

`local_operator/session/naming.py:303-304` defines `_cut_on_a_word = cut_on_a_word` "so the store below reads as one operation", but the alias is textually longer than the public name and has one call site. Using `cut_on_a_word` directly reads the same and removes a second name for one function.

---

#### What I verified and found correct

- **Rebase integrity.** No duplication of or conflict with main's `_on_conversation_name_written` (session.py:1418) or `_NAME_PERSIST_MAX_PASSES` (session.py:126). The shared-deadline change sits cleanly inside main's existing bounded-retry comment.
- **The dropped `_conversation_name_journalled` rollback is coherent, not a lost fix.** Main has no such field, and it has no `_conversation_name_journalled == …` guard for it to protect — main's flush gates on `_conversation_name_dirty` alone, and `_persist_conversation_name` clears that flag only after a successful append. So the transient-failure loss #166's F12 fixed is structurally impossible here. Confirmed by driving it: first append raises `OSError`, dispose retries, `appends attempted: 2`, one row on disk, `resumed name: 'Recovered'`. Correctly omitted.
- **No duplicate journal rows** at append 0.2/0.45/0.8 s against a 0.5 s budget — 1 row each.
- **Flush boundaries** (budget 0.5 s): 0.30/0.49/0.50 s land the title; 0.51/0.60 s are dropped and dispose returns at exactly 0.50 s. Zero and negative budgets degrade to an immediate `max(0.0, …)` return with the title dropped and warned — no hang, no negative timeout. The disclosed hang-then-raise case is indeed the only recovery path given up. 5.0 s is defensible: `dispose` already shields a 5.0 s turn-abort wait, so this matches the neighbouring bound rather than extending the worst case beyond it.
- **`cut_on_a_word`:** 20 000 randomized inputs x 6 caps produced **0** cap violations; a 200-char single token returns 79 chars + `…`; a fitting string is returned unchanged; caps 0/1/2 degrade to `'hello…'`/`'…'`/`'h…'` without raising. `parse_title`'s REJECT rule is unaffected — over-word and over-char model answers still return `None`.
- **The tui -> session.naming import adds nothing.** `local_operator.tui.settings` already pulls `local_operator.session.naming` transitively: importing `terminal_title` loads 266 modules vs 265 for `settings` alone, the one being `terminal_title` itself. All 9 `tests/unit/test_import_graph.py` guards pass.
- **Band invariants at HEAD on the swept full band** (4 ladder variants x alarm on/off x widths 60-180 x 9 titles incl. CJK/Hangul): 0 over-width rows, 0 non-flush rows, 0 orphaned seams, 0 alarm-column drift.
- **`_name_slack` staleness:** it is reset at the top of `_right_text` and read in the same call's join loop, so the `_fit` double-walk and the `forgo_name` restart cannot leave a stale value — the last `_right_text` call is always the one that painted the returned row.

Worktree left exactly at `42e66a6`; all 112 tests in the three touched files pass on the unmodified head.

Blockers to clear before this can merge: **F1**, plus **F2/F3** which fall out of it. Re-review the band hunk on the fixed head.

--- a/local_operator/session/naming.py
+++ b/local_operator/session/naming.py
@@ -301,10 +301,6 @@ def cut_on_a_word(text: str, max_chars: int) -> str:
     return cut.rstrip(" " + _TRAILING_PUNCTUATION) + "…"
 
 
-#: Module-private alias, so the store below reads as one operation.
-_cut_on_a_word = cut_on_a_word
-
-
 def _sentence_case(words: list[str]) -> str:
     """Capitalise the first word, leave every other word's casing alone.
 
@@ -522,7 +518,7 @@ class ConversationName:
         conversation name — every title reaching it had already been sliced by
         this line (design review round 2, D6).
         """
-        cleaned = _cut_on_a_word(" ".join((text or "").split()), MAX_TITLE_CHARS)
+        cleaned = cut_on_a_word(" ".join((text or "").split()), MAX_TITLE_CHARS)
         if not user_set and self.user_set:
             return self.text
         self.text = cleaned

--- a/local_operator/tui/widgets/status_line.py
+++ b/local_operator/tui/widgets/status_line.py
@@ -1506,9 +1506,14 @@ class StatusLine:
         """agents ‹ jobs ‹ context ‹ cost ‹ duration ‹ [!] ‹ name.
 
         ``name_cells`` is the box :meth:`_walk` reserved for the trailing name —
-        zero when the ladder dropped it. The name is PADDED to that box rather
-        than measured, which is what pins every other segment's column; see
-        :data:`NAME_CELLS`.
+        zero when the ladder dropped it. The row always SPENDS that whole box,
+        whatever the title's own width, which is what pins every other segment's
+        column; see :data:`NAME_CELLS`. The title itself is emitted UNPADDED and
+        the remainder (:attr:`_name_slack`) is paid out as blanks just before
+        the seam that introduces it, so the box is filled without the title
+        having to be the thing that fills it — which is what lets the row reach
+        the band's right edge at every title length while the columns to the
+        left of the name stay put.
 
         Colour groups by KIND rather than giving every field its own hue, which
         would be a rainbow: counters share `label`, the two numbers an operator
@@ -1733,22 +1738,24 @@ class StatusLine:
         that exactly fitted its frame `_MIN_GROUP_GAP` cells past it, on trailing
         blanks that aligned nothing.
 
-        The right group is aligned by its BOX, not by its ink: the name arrives
-        padded to the cells :meth:`_walk` reserved (:data:`NAME_CELLS`), so this
-        arithmetic is the same at every title length and a short title leaves its
-        leftover at the band's edge rather than shunting the group right.
+        The right group is aligned by its BOX, not by its ink: the group always
+        spends the cells :meth:`_walk` reserved for the name
+        (:data:`NAME_CELLS`), so this arithmetic is the same at every title
+        length and a short title cannot shunt the group right.
 
-        The box's unused cells are spent INSIDE the name segment, which
-        right-aligns itself in them (see :meth:`_right_text`), so the row's last
-        inked cell is the band's right edge at every title length and there is
-        nothing here to trim.
+        The box's unused cells are spent inside the right group by
+        :meth:`_right_text`, as a run of blanks immediately BEFORE the seam that
+        introduces the title — so the row's last inked cell is the band's right
+        edge at every title length and there is nothing here to trim.
 
-        This method used to `rstrip` a LEFT-aligned name's trailing padding,
+        This method used to `rstrip` a left-aligned name's trailing padding,
         which produced a row that ended early — flush for a long title and up to
-        35 cells short for a brief one. Moving the padding to the other side of
-        the title fixes that at the source: the cells still exist, so the fit
-        arithmetic and every column left of the name are untouched, but they are
-        no longer at the edge where their absence was visible.
+        35 cells short for a brief one. Moving those cells ahead of the title
+        fixes that at the source: they still exist, so the fit arithmetic and
+        every column left of the name are untouched, but they are no longer at
+        the edge where their absence was visible. Putting them between the seam
+        and the title instead reaches the edge too and orphans the chevron by up
+        to 39 cells, which reads as a segment that failed to render.
 
         Aligning the group by its INK rather than its box is the other way to
         reach the edge, and the one this must not do: measured across four

--- a/local_operator/tui/widgets/status_line.py
+++ b/local_operator/tui/widgets/status_line.py
@@ -1701,15 +1701,24 @@ class StatusLine:
             )
 
         right = Text()
+        last = len(parts) - 1
         for index, (icon, text, style) in enumerate(parts):
+            # The name's unused reserve is paid out HERE, immediately before the
+            # seam that introduces the title, so the chevron stays tight against
+            # the text it points at (design review D1) while the cells still sit
+            # inside the right group and cannot move any column to their left.
+            #
+            # OUTSIDE the `if index:` seam guard, because the name is not always
+            # preceded by a sibling: a freshly-opened session has no counters, no
+            # context reading, no cost and no duration, so the title is the whole
+            # right group and lands at index 0. Emitting the slack only alongside
+            # a seam silently skipped it there, and the group then aligned by its
+            # INK — the alternative this design explicitly rejects. Measured at
+            # 100 cells, the title's first column walked 95 -> 78 -> 73 -> 63 as
+            # the title grew, on the very first band a user sees (review F1).
+            if index == last and self._name_slack:
+                right.append(" " * self._name_slack, style=dim)
             if index:
-                # The name's unused reserve is paid out HERE, immediately before
-                # the seam that introduces the title, so the chevron stays tight
-                # against the text it points at (design review D1) while the
-                # cells themselves still sit inside the right group and cannot
-                # move any column to its left.
-                if index == len(parts) - 1 and self._name_slack:
-                    right.append(" " * self._name_slack, style=dim)
                 right.append(f" {_SEP_RIGHT} ", style=seam)
             if icon:
                 right.append(f"{icon} ", style=dim)

--- a/tests/unit/session/test_conversation_name_persistence.py
+++ b/tests/unit/session/test_conversation_name_persistence.py
@@ -287,13 +287,18 @@ async def test_a_slow_but_real_write_keeps_its_title(tmp_path, monkeypatch) -> N
     under the sum was cut off by the first half. One shared deadline keeps any
     append inside the budget and still bounds teardown.
     """
-    monkeypatch.setattr(session_mod, "_NAME_FLUSH_TIMEOUT_S", 0.6)
+    # NOT monkeypatched. The defect was a hard-coded per-half budget, so a test
+    # that patches the shared constant proves nothing about it: the patched name
+    # simply is not read by the broken code, and the test passes either way
+    # (review round 1, F4). The delay is therefore expressed against the REAL
+    # `_NAME_FLUSH_TIMEOUT_S` and placed where the two shapes disagree — past
+    # what a single half used to allow, comfortably inside the shared budget.
+    delay = session_mod._NAME_FLUSH_TIMEOUT_S * 0.6
+    assert delay > 2.0, "the delay must exceed the old per-half budget to discriminate"
 
     class SlowTranscript(Transcript):
         async def append_custom(self, custom_type: str, details: dict[str, Any]):
-            # Inside the shared budget, but past what either half would have
-            # allowed on its own.
-            await asyncio.sleep(0.45)
+            await asyncio.sleep(delay)
             return await super().append_custom(custom_type, details)
 
     session = Session(

--- a/tests/unit/tui/test_status_line.py
+++ b/tests/unit/tui/test_status_line.py
@@ -1379,6 +1379,15 @@ def test_a_brief_title_leaves_no_dead_run_at_the_bands_right_edge(monkeypatch) -
                 assert (
                     row == row.rstrip()
                 ), f"{where}: {len(row) - len(row.rstrip())} dead cells at the band's edge"
+                # THE property, and the one the assertion above cannot carry:
+                # the row must REACH the band's right edge, not merely end on a
+                # non-blank cell. A row whose trailing padding has been stripped
+                # satisfies `row == row.rstrip()` trivially while stopping up to
+                # 35 cells short — which is exactly what shipped, and why this
+                # test passed against the code it was written to pin (F2).
+                assert (
+                    cell_len(row) == width
+                ), f"{where}: the row stops {width - cell_len(row)} cells short of the edge"
                 # The ink really is the title's, not a coincidence of truncation.
                 assert row.endswith(truncate_name(name, _name_box(row, name, width))), where
                 boxes.add(_name_box(row, name, width))
@@ -1414,6 +1423,16 @@ def test_the_titles_seam_stays_tight_against_the_title(monkeypatch) -> None:
                 seam = f" {_SEP_RIGHT} "
                 tail = row[row.rindex(seam) + len(seam) :]
                 assert tail and not tail.startswith(" "), f"{where}: the seam was orphaned"
+                # Both halves together, because either alone is satisfied by a
+                # shape this test exists to reject: `ljust` + `rstrip` also
+                # leaves the seam tight (it just ends the row early), and
+                # `rjust` also reaches the edge (it just orphans the seam). Only
+                # asserting the pair pins the one arrangement that does both,
+                # and without this the test passed against the code it replaced
+                # (F2).
+                assert (
+                    cell_len(row) == width
+                ), f"{where}: the seam is tight but the row stops short of the edge"
 
 
 def test_a_double_width_title_never_overflows_the_band(monkeypatch) -> None:
@@ -1446,6 +1465,55 @@ def test_a_double_width_title_never_overflows_the_band(monkeypatch) -> None:
                         f"{label} alarm={alarm} width={width} name={name!r}: "
                         f"the row is {cell_len(row)} cells wide"
                     )
+                    if not status.is_showing("name"):
+                        # No name means no reserved box to fill, so the row is
+                        # whatever the surviving segments are — only the
+                        # overflow half of the contract applies here.
+                        continue
+                    # EQUAL, not merely `<=`. Both halves of the cell/character
+                    # mismatch are failures and only one of them is an overflow:
+                    # padding a double-width title by CHARACTERS overshoots the
+                    # box, and a row that then strips the excess lands SHORT of
+                    # the edge instead of over it (268 short rows measured that
+                    # way). Asserting the row is exactly its width catches the
+                    # mismatch in whichever direction it manifests — `<=` alone
+                    # passed against the very code this pins (F2).
+                    assert cell_len(row) == width, (
+                        f"{label} alarm={alarm} width={width} name={name!r}: "
+                        f"the row is {cell_len(row)} cells wide, not {width}"
+                    )
+
+
+def test_the_reserve_is_paid_even_when_the_name_is_the_only_segment(monkeypatch) -> None:
+    """The name is not always preceded by a sibling, and the slack must still land.
+
+    A freshly-opened session has no counters, no context reading, no cost and no
+    duration — ``format_jobs(0)``, ``format_agents(0)`` and
+    ``format_context_usage(0, …)`` all return "" and the duration is suppressed
+    at zero — so the title is the WHOLE right group and lands at index 0 of the
+    join loop. Emitting the reserve's slack only alongside a seam skipped it in
+    exactly that shape, and the group then aligned by its INK: the row stopped
+    short of the band's edge and the title's first column walked with its own
+    length, on the very first band a user sees (review round 1, F1).
+
+    This is the sparse band on purpose. Every other band test populates the row.
+    """
+    monkeypatch.setenv("HOME", "/Users/tester")
+    for width in (80, 100, 120, 160):
+        for name in ("a", "Short", "Reduce agent RAM usage", "A much longer title than that"):
+            dock = _dock(width)
+            status = StatusLine(dock)
+            status.update(
+                model_label="anthropic/claude-opus-4",
+                cwd="/Users/tester/work/local-operator",
+                conversation_name=name,
+            )
+            row = status.render_text(width).plain
+            if not status.is_showing("name"):
+                continue
+            assert (
+                cell_len(row) == width
+            ), f"width={width} name={name!r}: the row stops {width - cell_len(row)} cells short"
 
 
 def test_the_bands_cut_lands_on_a_word_like_the_excerpt_it_was_handed() -> None:

